### ISSUE TITLE
[QUESTION] Combining WebRtc SDK, Producer SDK and KVS PIC: what versions/tags to use?

### ISSUE THREAD

**Issue by roxlu** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#1385]
Maybe this is the wrong way to go about this, but I want to use the WebRtc SDK to setup a single low-latency video stream but also handle many viewers where higher latency is OK. A bit of background about what I'm working on: I send video from a device using WebRtc for a low-latency stream. Then I want to use the [Kinesis Producer SDK](https://github.com/awslabs/amazon-kinesis-video-streams-producer-c) as a secondary system which pushes video from this same device to AWS Kinesis that can then scale up to an unknown amount of viewers where a higher latency is OK. The WebRtc stream is working well using this repository. Now I'm looking into the Producer SDK.

The setup I've got in mind looks like:

```

   +------------+   +------------+   +------------+                   
   |            |   |            |   |            |                   
   |  viewer 1  |   |  viewer 2  |   |  viewer #  |                   
   |            |   |            |   |            |                   
   +------------+   +------------+   +------------+                   
             ^           ^              ^                                  
             |           |              |                                  
         +------------------------------+                                
         |                              |                                
         |      AVW Kinesis             |                                
         |                              |                                
         +------------------------------+                                
                       ^                                           
                       |                                           
                       |                                           
              +--------|-------+       +-----------------+         
              |                |       |                 |         
              |  video sender  |------>|  webrtc-viewer  |         
              |                |       |                 |         
              +----------------+       +-----------------+         

```

_I'm showing this flow as I'm curious if this might not be the best solution for my goals_.

Now, my question is, what versions of the dependent AWS repositories I should use? I noticed that this WebRtc repository depends on KVS PIC (which is also mentioned here #1363). But the PIC repository is unversioned. I think that this WebRtc repository and the Producer SDK both pull the correct PIC version?

When I add the producer sdk as an dependency I run into this error: 
`include/com/amazonaws/kinesis/video/webrtcclient/Include.h:1195:5: error: unknown type name ‘KvsRetryStrategyCallbacks’`

I'm currently using the 1.7.0 release of the WebRtc SDK.

**Comment by MushMal** [comment]
There is a good sample that is doing what you are attempting to do:

https://github.com/aws-samples/amazon-kinesis-video-streams-demos/tree/master/gst/gst-kvs-plugin

**Comment by roxlu** [comment]
Thanks @MushMal that's great! 

Though the issue I'm running into, is that `KvsRetryStrategyCallbacks`  was unknown. The .[gitmodules](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/v1.7.0/.gitmodules) of this repository pointed me into the wrong direction. After digging deeper into this it turns out that the .gitmodules is most likely [a left-over](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/commit/2f7db86a0be3cf341e5bce7e039e22ecf8dabedf) and can be removed, or at least for the `amazon-kinesis-video-streams-pic` repository; I didn't investigate this thoroughly. 

The amazon-kinesis-video-streams-pic is now pulled in through the amazon-kinesis-video-streams-producer-c via this [external project](https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/master/CMake/Dependencies/libkvspic-CMakeLists.txt) description. And the `amazon-kinesis-video-streams-producer-c` is pulled through [this external project](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/v1.7.0/CMake/Dependencies/libkvsCommonLws-CMakeLists.txt). Though this repository does provide the `libcproducer.a`  from the `amazon-kinesis-video-streams-producer-c`.  

About the link you sent: I'm going to experiment to deliver the `Frame` instances I already have in my current code, to a STREAM_HANDLE using the `putKinesisVideoFrame()` function. I expect that to work, but haven't tested with this yet.

**Comment by MushMal** [comment]
WebRtc would only need the Utility sub-project from the PIC. As the producer includes the entire PIC the webrtc solution can simply use the PIC as is. There is another shared piece - the common library from C producer. Similar scenario with the transitive closure. Make sure you set the compile flags accordingly - for example, to use curl vs libwebsockets.

The link I sent which contains the GStreamer plugin that works for both WebRTC and as a KVS producer does exactly that. It continuously streams into KVS but once it has a WebRTC viewer, it would also produce the same frame into the transceiver.

**Comment by roxlu** [comment]
Thanks again @MushMal I tried to create a similar CMakelists.txt like [this](https://github.com/aws-samples/amazon-kinesis-video-streams-demos/blob/master/gst/gst-kvs-plugin/CMakeLists.txt) but instead of using the GIT_TAGs I used the master branch as the versions used in that plugin are pretty old. This results in several linker issues. As each repository builds several libraries and both the webrtc and producer it takes a lot of time digging into the build system to figure out what symbol is defined in what library.  It would be great if these libraries were a bit more in sync and more easily used together.

I'm going to look into another solution for now but might dive back into this.
